compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/draganddrop/DragAndDrop.desktop.kt

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.draganddrop
+
+import androidx.compose.ui.DragData
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.dragData
+import androidx.compose.ui.geometry.Offset
+import java.awt.datatransfer.Transferable
+import androidx.compose.ui.platform.PlatformDragAndDropManager
+import java.awt.dnd.DropTargetDropEvent
+
+/**
+ * Encapsulates the information needed to start a drag-and-drop session from Compose on the desktop.
+ */
+@OptIn(ExperimentalComposeUiApi::class)
+actual class DragAndDropTransferData(
+    val transferable: DragAndDropTransferable,
+    val supportedActions: Iterable<DragAndDropTransferAction>,
+    val initialAction: DragAndDropTransferAction,
+    val dragOffset: Offset = Offset.Zero,
+    val onTransferCompleted: ((userAction: DragAndDropTransferAction?) -> Unit)? = null,
+)
+
+/**
+ * Represents the actual object transferred during a drag-and-drop.
+ */
+@ExperimentalComposeUiApi
+abstract class DragAndDropTransferable {
+    /**
+     * Returns the AWT [Transferable] to pass to the AWT drag-and-drop system.
+     *
+     * This must return a non-null value in order to work with the AWT implementation of
+     * [PlatformDragAndDropManager].
+     */
+    abstract fun toAwtTransferable(): Transferable?
+}
```

Comment:
```suggestion
abstract class DragAndDropNativeTransferable
```

and:
```
@ExperimentalComposeUiApi
val DragAndDropNativeTransferable.awtTransferable: Transferable
    get() = (this as DropTargetDropEvent).transferable
```

Otherwise it won't be possible to create a custom platform implementation.

It doesn't have type-safety, but we can achieve it only if we change the architectural approach to "one-platform-in-runtime" (and provide ability to write own sourcesets) instead of "any-plaftorm-in-runtime", which we need to discuss separately.

Reply:
It's possible to create a custom platform implementation - it just needs to return `null` in `toAwtTransferable`.

What you're suggesting will require an additional abstract class:
```
abstract class AwtDragAndDropTransferable: DragAndDropNativeTransferable() {
    abstract fun toAwtTransferable(): Transferable
}
```
and this is even more confusing.

Reply:
>  it just needs to return null in toAwtTransferable.

We shouldn't have opinionated platform class in any non-optional Compose core API. `toAwtTransferable` is opinionated and currently it is required to be implemented by other platforms. If we say that it is possible to implement other platforms with `nil` value, then we should apply the same approach for our AWT implementation. Instead of member `toAwtTransferable`  we should add an extension.

> What you're suggesting will require an additional abstract class:

We don't need an additional class with this code:
```
@ExperimentalComposeUiApi
class AwtDragAndDropTransferable(
    val transferable: Transferable
) : NativeDragAndDropTransferable()

@ExperimentalComposeUiApi
val NativeDragAndDropTransferable.awtTransferable: Transferable
    get() = (this as AwtDragAndDropTransferable).transferable
```

Reply:
AwtDragAndDropTransferable (the current version) is just one possible implementation - one where the user provides a Transferable directly. As we discussed, we would want to add additional classes (e.g. TextDragAndDropTransferrable), and they will all need a `toAwtTransferrable` method. So they will need a base type with a declaration of that method.

Reply:
> we would want to add additional classes (e.g. TextDragAndDropTransferrable)

Yes, we will need them in the future. But they should work with any platform implementation, so it isn't `TextDragAndDropTransferrable` that needs to provide `toAwtTransferrable`. It is something else should map simple "data" class `TextDragAndDropTransferrable` to `awtTransferable`. For example we can map it at the call site inside `AwtDragAndDropManager`.

Reply:
That would mean we'd have to check the type with `is`, and we'd be limited to the set of known types. Users would not be able to implement their own.

Reply:
It seems that we don't need the current `toAwtTransferrable` or my suggested `awtTransferable` public at all. It is only used by our internal `DragAndDropManager` implementation. A custom implementation should use their own type.

The main requirements are to be able to write this:
```
ComposeScene(
  platformContext = object : PlatformContext {
    override fun createDragAndDropManger() = myStrangePlatformDragAndDropManager()
  }
).apply {
  setContent {
    DragAndDropTransferData(
       transferable = TextDragAndDropTransferrable(exportedText),
       ...
    )

    DragAndDropTransferData(
       transferable = MyStrangePlatformTransferrable(strangeItem),
       ...
    )
  }
}
```

And this:
```
singleWindowApplication {
  DragAndDropTransferData(
     transferable = AwtDragAndDropTransferable(
        StringSelection(exportedText)
     ),
     ...
}
```

Reply:
Still, if we don't put an abstract `toAwtTransferable` method in some base type, we limit the set of types to the ones `AwtDragAndDropManager` knows about. Allowing users to implement their own `DragAndDropTransferrable` is the point of that method.

Reply:
> Allowing users to implement their own DragAndDropTransferrable is the point of that method.

With the current API they can only implement it for AWT platform. If that is what they needed, they can create it this way:
```
fun TextDragAndDropTransferrable(exportedText: String): DragAndDropTransferable = 
    AwtDragAndDropTransferable(StringSelection(exportedText))
```